File: leetcode/201-300/214-shortest-palindrome/bench/shortest_palindrome.c
#include <stdio.h>
#include <stdlib.h>

/* Benchmark workload for LeetCode #214 — Shortest Palindrome.
 *
 * The kata's core is the KMP prefix-function over comb = s + '#' + reverse(s);
 * fail[last] is the longest-palindromic-prefix (lps) length. One call is O(n).
 * The bench turns it into a throughput kernel: build ONE big symbol string over
 * a tiny alphabet (so palindromic prefixes actually occur and lps varies), then
 * slide a fixed-width window across it and, for every window, build comb and run
 * the prefix-function, summing lps. The prefix-function's `len = fail[len-1]`
 * fallback is a loop-carried, data-dependent recurrence — it does NOT vectorize.
 * Sink = sum of lps over all windows.  Symbols are i64 codes (alphabet 0..A-1),
 * with -1 as the '#' separator, so all five languages run the identical kernel. */

#define BIG    260000L   /* length of the base symbol string          */
#define W      512L      /* window width                               */
#define ALPHA  2L        /* alphabet size (small => rich palindromes)  */

int main(void) {
    long *base = malloc(BIG * sizeof(long));
    long state = 12345;
    for (long i = 0; i < BIG; i++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        base[i] = state % ALPHA;
    }

    long m = 2 * W + 1;                 /* comb length: W + 1 + W */
    long *comb = malloc(m * sizeof(long));
    long *fail = malloc(m * sizeof(long));

    long windows = BIG - W;             /* stride 1 sliding windows */
    long sink = 0;
    for (long w = 0; w < windows; w++) {
        /* comb = base[w..w+W] + '#' + reverse(base[w..w+W]) */
        for (long i = 0; i < W; i++) comb[i] = base[w + i];
        comb[W] = -1;
        for (long i = 0; i < W; i++) comb[W + 1 + i] = base[w + W - 1 - i];

        fail[0] = 0;
        long len = 0;
        long idx = 1;
        while (idx < m) {
            if (comb[idx] == comb[len]) {
                len += 1;
                fail[idx] = len;
                idx += 1;
            } else if (len > 0) {
                len = fail[len - 1];
            } else {
                fail[idx] = 0;
                idx += 1;
            }
        }
        sink += fail[m - 1];
    }

    printf("%ld\n", sink);
    free(base); free(comb); free(fail);
    return 0;
}

File: leetcode/201-300/214-shortest-palindrome/bench/shortest_palindrome.rs
fn prefix_lps(base: &[i64], w: i64, width: i64, comb: &mut [i64], fail: &mut [i64]) -> i64 {
    let m = 2 * width + 1;
    for i in 0..width { comb[i as usize] = base[(w + i) as usize]; }
    comb[width as usize] = -1;
    for i in 0..width { comb[(width + 1 + i) as usize] = base[(w + width - 1 - i) as usize]; }

    fail[0] = 0;
    let mut len = 0i64;
    let mut idx = 1i64;
    while idx < m {
        if comb[idx as usize] == comb[len as usize] {
            len += 1;
            fail[idx as usize] = len;
            idx += 1;
        } else if len > 0 {
            len = fail[(len - 1) as usize];
        } else {
            fail[idx as usize] = 0;
            idx += 1;
        }
    }
    fail[(m - 1) as usize]
}

fn main() {
    let big: i64 = 260000;
    let width: i64 = 512;
    let alpha: i64 = 2;

    let mut base: Vec<i64> = Vec::new();
    let mut state: i64 = 12345;
    for _ in 0..big {
        state = (state * 1103515245 + 12345) & 2147483647;
        base.push(state % alpha);
    }

    let m = 2 * width + 1;
    let mut comb: Vec<i64> = vec![0; m as usize];
    let mut fail: Vec<i64> = vec![0; m as usize];

    let windows = big - width;
    let mut sink: i64 = 0;
    for w in 0..windows {
        sink += prefix_lps(&base, w, width, &mut comb, &mut fail);
    }
    println!("{}", sink);
}
